backend/src/main/java/gov/cdc/usds/simplereport/config/authorization/DemoAuthenticationConfiguration.java

```diff
@@ -0,0 +1,153 @@
+   *
+   * @param config the configured list of DemoUsers, with or without a default user set.
+   */
+  private static Supplier<Optional<DemoUser>> getCurrentDemoUserSupplier(
+      DemoUserConfiguration config) {
+    return () -> {
+      Authentication auth = SecurityContextHolder.getContext().getAuthentication();
+      String userName = null;
+      if (auth == null) { // have to allow defaulting to work even when there is no authentication
+        LOG.warn("No authentication found: current user will be default (if available)");
+      } else if (auth.isAuthenticated()) {
+        userName = auth.getName();
+      } else {
+        LOG.info("Unauthenticated user in demo mode: current user will be default (if available)");
+      }
+      DemoUser found = config.getByUsername(userName);
+      if (found == null) {
+        LOG.error("Invalid user {} in demo mode", userName);
+      }
+      return Optional.ofNullable(found);
+    };
+  }
+
+  /**
+   * An {@link AuthorizationService} that looks up the username of the current authenticated user in
+   * the list of configured demo users, falling back to the default if one is configured, and
+   * returns the claims for that {@link DemoUser}. See {@link DemoUserIdentitySupplier} for snarky
+   * comments about this idea.
+   */
+  public static class DemoUserAuthorizationService implements AuthorizationService {
+
+    private final Supplier<Optional<DemoUser>> _getCurrentUser;
+
+    public DemoUserAuthorizationService(DemoUserConfiguration config) {
+      _getCurrentUser = getCurrentDemoUserSupplier(config);
+    }
+
+    public DemoUserAuthorizationService(List<DemoUser> users) {
+      this(new DemoUserConfiguration(users));
+    }
+
+    @Override
+    public List<OrganizationRoleClaims> findAllOrganizationRoles() {
+      return _getCurrentUser.get().map(DemoUser::getAuthorization).map(List::of).orElse(List.of());
+    }
+  }
+
+  /**
+   * An {@link IdentitySupplier} that looks up the username of the current authenticated (or, more
+   * likely, "authenticated") user in the list of configured demo users, falling back to the default
+   * if one is configured. It is like that had we set out to have this originally we would have
+   * ended up with a more conventional UserDetailsService, and we could eventually work our way back
+   * there, but path-dependence produces weird results sometimes.
+   */
+  public static class DemoUserIdentitySupplier implements IdentitySupplier {
+
+    private final Supplier<Optional<DemoUser>> _getCurrentUser;
+
+    public DemoUserIdentitySupplier(DemoUserConfiguration config) {
+      this._getCurrentUser = getCurrentDemoUserSupplier(config);
+    }
+
+    public DemoUserIdentitySupplier(List<DemoUser> wiredUsers) {
+      this(new DemoUserConfiguration(wiredUsers));
+    }
+
+    @Override
+    public IdentityAttributes get() {
+      return _getCurrentUser.get().map(DemoUser::getIdentity).orElse(null);
```

Comment:
i don't support this idea of returning a null Identity -- i think we should throw an Exception before we ever do that.

Reply:
We return null from the live implementation, and check for it in ApiUserService.getCurrentUser to see if we need to look for a patient identity instead—I don't disagree that it's a weird thing to do, and we could change the API to return an Optional instead, but it's not an exceptional case so I don't think we should throw an exception.